### DISCUSSION TITLE
Hard Coded Context Length ?

### DISCUSSION THREAD

**Discussion by Tum1370** [ace-step/ACE-Step-1.5#279] [General]
Does it have a hard coded context length for the Model to work with ??

Is this why when setting a 200 second duration length, still runs out of time and stops a lyrics abruptly (maybe a 3 words) before the end of the song.
If you then increase the duration to 204 seconds. It still misses the last few words. 
It feels like the Model Context is totally running out by the end of the song ??

**Comment by sigalarm** [comment]
Another thing that may need to be added to an eventual FAQ, some information from the code base:
---------------------

# FAQ: Context Length and Lyrics Cutoff on Long Duration Songs

## Question

> Does it have a hard coded context length for the Model to work with?
>
> Is this why when setting a 200 second duration length, still runs out of time and stops lyrics abruptly (maybe 3 words) before the end of the song. If you then increase the duration to 204 seconds, it still misses the last few words. It feels like the Model Context is totally running out by the end of the song.

## Answer

**Yes, there is a hard-coded context length, and yes, it's likely what's causing the lyrics cutoff.**

Here's the technical breakdown:

### The 4096 Token Context Window

The 5Hz Language Model (LM) has a hard-coded `max_model_len = 4096` tokens (`llm_inference.py:41`, also enforced at line 464). On low-VRAM GPUs (<8GB), this drops to **2048** tokens (line 462).

This is the **total** context window — it must fit **both** the input prompt AND the generated output.

### How Duration Maps to Tokens

The model is called "5Hz" because it generates **5 audio semantic tokens per second**. So:

| Duration | Audio Tokens Needed |
|----------|-------------------|
| 60s | 300 tokens |
| 120s | 600 tokens |
| 200s | 1,000 tokens |
| 204s | 1,020 tokens |

### The Budget Breakdown

The code at `llm_inference.py:542-552` calculates the token budget like this:

```
max_tokens = min(duration * 5 + 500, max_model_len - 64)
```

So for a 200s song:
- **Audio codes needed:** 200 x 5 = **1,000 tokens**
- **CoT/metadata overhead:** ~**500 tokens** (for the `<think>` block with BPM, key, caption, etc.)
- **Total output needed:** ~**1,500 tokens**
- **Input prompt** (lyrics, caption, formatting): this consumes maybe **200-500+ tokens** depending on lyric length
- **Max output allowed:** 4096 - 64 = **4,032 tokens**

At first glance, 4,032 seems like enough for 1,500 tokens of output. But here's the catch — this is the **total sequence length** including the input prompt. The vLLM engine treats `max_model_len` as the combined input + output limit.

### Why Lyrics Cut Off

The **constrained logits processor** (`constrained_logits_processor.py:1594-1607`) enforces the target duration by:
1. **Blocking EOS** until exactly `duration * 5` audio codes are generated
2. **Forcing EOS** once the target count is reached

But if the model hits the `max_model_len` ceiling before generating all the required audio codes, **generation simply stops** — the LM runs out of context space. The audio codes produced so far get decoded, and you get a song that's slightly shorter than requested, with lyrics abruptly cut off near the end.

Increasing from 200s to 204s doesn't help because the problem isn't a few tokens short — it's that the context window is being exhausted by the combination of prompt + metadata + audio codes.

### The Real Limits

Roughly speaking, with `max_model_len = 4096`:
- **Practical max duration** with lyrics + CoT: ~600-700 seconds of audio codes is theoretically possible, but with a long lyrics prompt, the real ceiling is lower
- Longer/more complex lyrics = more input tokens = fewer tokens available for audio codes
- The CoT `<think>` block adds overhead for metadata reasoning

If you're on a **low-VRAM GPU** where `max_model_len` drops to **2048**, the maximum practical duration is roughly halved.

### Suggestions

1. **Check your GPU tier** — if you have <8GB VRAM, the context is only 2048, which severely limits long songs
2. **Shorter lyrics** for long songs — less input prompt means more room for audio codes
3. **Use instrumental mode** for very long durations — no lyrics = smaller prompt = more headroom
4. **The 5Hz LM is only the "planner"** — for very long songs, consider using DiT-only mode (`init_llm=false`) where the DiT handles generation directly without the LM context limitation

**Comment by Tum1370** [comment]
How can i use "init_llm=false" when am using ComfyUI and Ace-Step in the Pinokio setup ? 
Which file am i looking for ?

**Reply by sigalarm** [reply]
Try this - 
Set the environment variable ACESTEP_INIT_LLM=false — this works at the ACE-Step core level regardless of frontend (ComfyUI, Gradio, API, etc.).

**Reply by sigalarm** [reply]
A more detailed answer
----------

The ACE-Step ComfyUI nodes are a **separate third-party project** (not part of the core ACE-Step repo), so the exact node UI may vary. However, ACE-Step reads the LLM setting from an **environment variable**, which works regardless of which frontend you're using:

### Option 1: Set the Environment Variable (Recommended — Works Everywhere)

The core ACE-Step code checks the `ACESTEP_INIT_LLM` environment variable before deciding whether to load the LLM. Setting this will work whether you're using ComfyUI, Gradio, the REST API, or any other frontend.

**On Windows (Pinokio typically runs on Windows):**

Before launching Pinokio/ComfyUI, set the variable in a terminal:
```cmd
set ACESTEP_INIT_LLM=false
```

Or set it permanently via **System Settings > Environment Variables**.

**On Linux/Mac:**
```bash
export ACESTEP_INIT_LLM=false
```

**In a `.env` file:** If ACE-Step's install directory contains a `.env` file (or `.env.example` you can copy), add or change:
```
ACESTEP_INIT_LLM=false
```

The `.env.example` in the ACE-Step project documents all the valid values:
- `auto` (default) — let GPU detection decide
- `true` / `1` / `yes` — force enable LLM (may cause OOM on low VRAM)
- `false` / `0` / `no` — force disable LLM (pure DiT mode)

### Option 2: Look for the Setting in the ComfyUI Node

Many ACE-Step ComfyUI node implementations expose an `init_llm` toggle directly in the node's configuration panel. Look for:
- A checkbox or dropdown labeled **"Init LLM"**, **"Enable LLM"**, or **"Thinking Mode"**
- It may be on the loader/setup node rather than the generation node

If the ComfyUI nodes don't expose this setting, the environment variable approach (Option 1) will override it at the ACE-Step core level regardless.

### What You Lose in DiT-Only Mode

When the LLM is disabled, these features become unavailable:
- **Thinking mode** (Chain-of-Thought reasoning for BPM, key, caption)
- **Sample mode** (generate music from a natural language description)
- **Format mode** (LLM-enhanced lyrics formatting)
- Automatic language/genre detection via the LM

The DiT model still handles the actual audio synthesis and can produce full-length songs — it just won't have the LM's "planning" step, which means you may need to be more explicit with your metadata (BPM, key, etc.) and the model relies on its own internal text encoder for lyric alignment rather than the 5Hz LM's semantic tokens.

### Pinokio-Specific Tips

Pinokio manages apps through scripts in its app directory. Look for:
- A `start.sh` or `start.bat` script in the ACE-Step app folder
- A `pinokio.js` or configuration file where environment variables are set
- You can add `ACESTEP_INIT_LLM=false` to whichever script launches the ACE-Step service

**Reply by Tum1370** [reply]
Thank you, i will give it a try. But I am still afraid it will feel like a downgrade from the previous version.
I have been trying everything with this for a  couple of full days now and cannot produce anything decent.
The idea of the lyric tags is very nice, but am guessing that wont be there with this setting.
I do not care about turbo speed for generating, I value quality over speed.

**Comment by Tum1370** [comment]
This version of 1.5 seems a huge downgrade in my opinion. The version 1 was producing far better songs / music.
Even if i did get that setting you mentioned above, it would probably then mean that the Lyric tags wouldnt work.

**Reply by ChuxiJ** [reply]
Don't draw conclusions when you're not using it correctly.
And don't draw conclusions when you haven't read the docs to understand how it works.

**Reply by Tum1370** [reply]
> Don't draw conclusions when you're not using it correctly. And don't draw conclusions when you haven't read the docs to understand how it works.

Its ok saying that. I have just tested both the comfyui version and the Gradio version. No matter how simple i make my prompt,, i have all the tags in the lyrics, i have tested all the models on the gradio version, and still, it ignores / misses words, whole lines. It just doesnt follow the lyrics however many times you try.
I have read your instruction page. And still no joy after a lot of testing.

No matter how strict you set it to follow everything, it does not follow. 
Also it doesnt even follow the simple setting of FLAC, and when you download the song its in mp3.

Am sorry, but i have tested this a lot now over the last 3 days, and cannot, not matter how many times trying, get it to work correctly. 
Yes the songs do sound ok, maybe some distortion but that can be sorted. But missing words / lines is just no good.

EDIT: I even stripped down the song to just 2 verses, pre chorus, 2 chorus, so and removed loads of lyrics as a test. Yet still it misses whole lines of words from the lyrics.   
These latest tests were all on the Gradio version.

Am trying to give you some constructive advice on my findings over the last 3 days.
